// jsp-raknet/js/protocol/NewIncomingConnection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Packet_1 = __importDefault(require("./Packet"));
const Identifiers = require('./Identifiers');
class NewIncomingConnection extends Packet_1.default {
    constructor() {
        super(Identifiers.NewIncomingConnection);
        this.systemAddresses = [];
    }
    decode() {
        super.decode();
        this.address = this.readAddress();
        // Do not save in memory stuff we will not use
        for (let i = 0; i < 20; i++) {
            this.systemAddresses.push(this.readAddress());
        }
        this.requestTimestamp = this.readLong();
        this.acceptedTimestamp = this.readLong();
    }
    encode() {
        super.encode();
        this.writeAddress(this.address);
        for (let address of this.systemAddresses) {
            this.writeAddress(address);
        }
        this.writeLong(this.requestTimestamp);
        this.writeLong(this.acceptedTimestamp);
    }
    static from(buffer) {
        const e = new NewIncomingConnection();
        e.buffer = buffer;
        e.decode();
        return e;
    }
}
exports.default = NewIncomingConnection;
